src/esp/assets/managers/AssetAttributesManager.h

```diff
@@ -163,46 +149,321 @@ class AssetAttributesManager
+      return nullptr;
+    }
+    return std::dynamic_pointer_cast<U>(res);
+  }  // AttributesManager::getTemplateCopyByHandle
 
  protected:
+  /**
+   * @brief Verify that passed template handle describes attributes of type
+   * specified by passed primtive name (ie "cube", "capsule")
+   * @param templateHandle The handle to test.
+   * @param attrType Type of primitive the attributes owning templateHandle
+   * describe.
+   * @return If template handle describes template for objects of desired
+   * primitive class name
+   */
+  bool verifyTemplateHandle(const std::string& templateHandle,
+                            const std::string& attrType) {
+    if (std::string::npos == templateHandle.find(attrType)) {
+      LOG(ERROR) << "AssetAttributesManager::verifyTemplateHandle : Handle : "
+                 << templateHandle << " is not of appropriate type for desired "
+                 << attrType << " primitives.  Aborting.";
+      return false;
+    }
+    return true;
+  }  // AttributesManager::verifyTemplateHandle
+
+  /**
+   * @brief Add an @ref AbstractPrimitiveAttributes object to the @ref
+   * templateLibrary_.
+   *
+   * @param attributesTemplate The attributes template.
+   * @param ignored Not used for asset attributes templates - handle is derived
+   * by configuration.
+   * @return The index in the @ref templateLibrary_ of object
+   * template.
+   */
+  int registerAttributesTemplateFinalize(
+      AbstractPrimitiveAttributes::ptr attributesTemplate,
+      const std::string& ignored = "") override;
+
+  /**
+   * @brief Whether template described by passed handle is read only, or can be
+   * deleted.  Default primitive asset templates should not be removed.
+   * @param templateHandle the handle to the template to verify removability.
+   * Assumes template exists.
+   * @return Whether the template is read-only or not
+   */
+  bool isTemplateReadOnly(const std::string& templateHandle) override {
+    for (auto handle : defaultTemplateNames) {
+      if (handle.compare(templateHandle) == 0) {
+        return true;
+      }
+    }
+    return false;
+  };
+
   /**
    * @brief Build an @ref AbstractPrimtiveAttributes object of type associated
    * with passed class name
+   * @param primClassName Magnum::Primitives class name of primitive being
+   * constructed
    */
-  const AbstractPrimitiveAttributes::ptr buildPrimAttributes(
-      const std::string& primTypeName) {
-    if (primTypeConstructorMap_.count(primTypeName) == 0) {
+  AbstractPrimitiveAttributes::ptr buildPrimAttributes(
+      const std::string& primClassName) {
+    if (primTypeConstructorMap_.count(primClassName) == 0) {
       LOG(ERROR) << "AssetAttributesManager::buildPrimAttributes : No "
-                    "primitive of type"
-                 << primTypeName << "exists.  Aborting.";
+                    "primitive class"
+                 << primClassName << "exists in Magnum::Primitives.  Aborting.";
       return nullptr;
```

Comment:
Maybe off-topic for this PR, but I want to ask about this error handling. Would we consider this a "fatal" error? If yes, then shouldn't we exit? If no, it doesn't seem like returning nullptr will be safe--I saw some calling code that seems to assume the return value will be non-null.

Reply:
There might be some situations that are not checking for nullptr, but all callers of this should be checking. I'll investigate.  We don't want to kill everything if this fails, though, since it only means they got the class name wrong.

Reply:
Found a few callers who were not checking.  Thanks for finding this.